src/hotspot/share/runtime/handshake.hpp

```diff
@@ -39,74 +40,75 @@ class JavaThread;
 // the target thread in a blocked state. A handshake can be performed with a
 // single JavaThread as well. In that case, the callback is executed either
 // by the target JavaThread itself or, depending on whether the operation is
-// a direct handshake or not, by the JavaThread that requested the handshake
-// or the VMThread respectively.
-class HandshakeClosure : public ThreadClosure {
+// a single target/direct handshake or not, by the JavaThread that requested the
+// handshake or the VMThread respectively.
+class HandshakeClosure : public ThreadClosure, public CHeapObj<mtThread> {
```

Comment:
Just to be clear. You haven't added support for a handshake that
must only be executed by the target thread yet, right? That's
future work, if I remember correctly...

Reply:
Yes as David says. The code that make sure this is case is:
static bool processor_filter(HandshakeOperation* op) {
  return !op->is_asynch();
}
Which is used e.g:
  return _queue.contains(processor_filter);

While process_self_inner() uses no filter.